feat: Improve error handling

ValuedMammal opened the pull request:
Changed some functions that return `Option` to return a `Result` instead, which makes for better error handling.

This is based on the work in #2 
    
Added
    
- `FromPsbtInputError`
- `CoinbaseMismatch`
- `CannotMeetTarget`
- `SelectorError`
- `GetForeignUnspentError`
- `ExtractReplacementsError`

I also fixed a few leftover review comments from [bitcoindevkit/bdk-tx#1](https://github.com/bitcoindevkit/bdk-tx/pull/1).

ValuedMammal committed:
feat!: Improve error handling

Change functions that return Option to return a Result instead
which makes for better error handling.

Added

- `FromPsbtInputError`
- `CoinbaseMismatch`
- `CannotMeetTarget`
- `SelectorError`
- `GetForeignUnspentError`
- `ExtractReplacementsError`

evanlinjin commented on src/input.rs:
```diff
@@ -181,33 +240,59 @@ impl Input {
+            _ => return Err(FromPsbtInputError::UtxoCheck),
+        };
         let prev_tx = psbt_input.non_witness_utxo.clone().map(Arc::new);
+        let is_coinbase = prev_tx
+            .as_ref()
+            .map(|tx| tx.is_coinbase())
+            .unwrap_or_default();
```
This won't work if we only have `witness_utxo`. Maybe it makes more sense to get the caller to pass in an `is_coinbase` parameter?

ValuedMammal replied:
I added the `is_coinbase` parameter to `Input::from_psbt_input` and `try_get_foreign_unspent` along with a new error type `CoinbaseMismatch`.

ValuedMammal committed:
ci: fix typo in rust.yml

ValuedMammal committed:
feat: Add `Selector::select_with_algorithm`

ValuedMammal commented:
Rebased onto fee4e672312f4c2ac5e9cbe8c76d1bc267131179

- Fixed a typo in rust.yml
- Renamed `CanonicalUnspentsError` to `GetForeignUnspentError`
- Added `CoinbaseMismatch` error
- Changed return type of `Selector::try_finalize` back to Option, as it seemed to make sense
- Added `Selector::select_with_algorithm`

ValuedMammal committed:
fix(finalizer): return true if already finalized in `finalize_input`

Include the check for `final_script_sig` when evaluating whether the
input is finalized.

evanlinjin commented on src/input.rs:
```diff
@@ -33,29 +36,49 @@ impl InputStatus {
+    /// match the actual satisfaction of the `psbt_input`.
     ///
-    /// # WHy do we only support finalized psbt inputs?
+    /// ## Why do we only support finalized psbt inputs?
     ///
     /// There is no mulit-party tx building protocol that requires choosing from foreign,
     /// non-finalized PSBT inputs.
```
I think it is best we change this to allow for non-finalized PSBT inputs since it makes our API more flexible.

evanlinjin reviewed (commented):
Looks great, just a find request.

Can address the TODOs later on.

evanlinjin reviewed (approved):
ACK 5bc533da6e251300249d37656fda1e276ee857bf
